core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import cats.implicits._
+import rankn._
+
+import Identifier.Constructor
+
+sealed abstract class NormalExpression {
+  val maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(Pattern[(PackageName, Constructor), Type], NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1).filter(_ >= 0)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression])
+  extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) {
+    val maxLambdaVar = None
+  }
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+}
+
+case class NormalizePackageMap(pm: PackageMap.Inferred) {
+  import Normalization._
+  import TypedExpr._
+
+  val normalizePackageMap: PackageMap.Normalized = 
+    PackageMap(
+      pm.toMap.toList.map { case (name, pkg) => {
+        normalizePackage(name, pkg).map((name, _))
+    }}.sequence.run(Map()).value._2.toMap)
+
+  def normalizeExpr(expr: TypedExpr[Declaration], env: Env, p: Package.Inferred): State[
+    Map[(PackageName, Identifier), ResultingRef],
+    TypedExpr[(Declaration, NormalExpressionTag)]] =
+      expr match {
+        case a@Annotation(_, _, _) => normalizeAnotation(a, env, p)
+        case g@Generic(_, _, _) => normalizeGeneric(g, env, p)
+        case v@Var(_, _, _, _) => normalizeVar(v, env, p)
+        case al@AnnotatedLambda(_, _, _, _) => normalizeAnnotatedLambda(al, env, p).map(_.asInstanceOf[TypedExpr[(Declaration, NormalExpressionTag)]])
+        case a@App(_, _, _, _) => normalizeApp(a, env, p).map(_.asInstanceOf[TypedExpr[(Declaration, NormalExpressionTag)]])
```

Comment:
I wonder how hard it would be to remove these casts. What are the blockers to that?

Reply:
There is a place where I call `normalizeApp` and I want to then deconstruct the resulting `App` to get out the normal expression so the resulting type is `TypedExpr.App` and not `TypedExpr`. I think there is a way I could fix the signature of `normalizeExpr` to accept extensions of `TypedExpr`. But I couldn't figure out the syntax.